### ISSUE TITLE
Type mismatch of `this` in an arrow function not reported as an error

### ISSUE THREAD

**Issue by sebastian-nowak** [microsoft/TypeScript#60188]
### 🔎 Search Terms

"arrow function", "this", "this inference", "nested arrow function"

### 🕗 Version & Regression Information

This is the behavior in every version I tried up to 5.6.3. Nightly is affected too.

### ⏯ Playground Link

https://www.typescriptlang.org/play/?ts=5.6.3#code/KYDwDg9gTgLgBDAnmYcBiBXAdgYwDwBQccAkllsFAIJQDmAznKDMFgCaPYDWWEA7lgDaAXQIA+OAF44AChgALAJb0AXOmz4yFanXpiAlFIkA3CIrYBuAgVCRYCZKjIAzSlso0GeACoTpMojhvJhAWdkZMXDxFLFcoUnIPXQkAfgTtT0Y1CmNKAn0rAmcNGEUILDgY7QBhAEMAG3q8dx0GELCOOG5eARExAOJnLDU5JVV1KPSkhjEAGjgAOiXa3TUWzMNJEzM2fLgAb0CAeiPFpYIAX2sccvp4IbVI-EEsDABbACNKYT84YtxSuUZIZDsQTpVYpQIho8C93l8oD9jqcFMpKowblAoMAcDBAqj6IUwacYnFGP9cWUKrUYBA3lRKeVYeZsvDvv0HrICY8YYIWXBXp92fN+YKEZttuZ9GpTOZkQgxui4JjsbjAlVKHVGjIZKK2VAJQdAsSIWS4HChYi4ABaZXQVV44jEJAoOArWjeRxSBJxda6PAu4AQZwK5RiKxOuDg0lQibPMXsm12rE4x1OgkRuAXArWE0xqDkkpUt20+mMrDMthqO5QGK0H4yTmjZQ8qJ8qtwGt1n4ijtdrC0Q2ytjSuDD+UAdQASgB5AByAHEk6isFxGAoaQrUCsoPw-lgAOSMAAGBOPDldaNqcDYimccVY8HK2-YN4gwHoh-g2Ls8FqFUoXcoHVRIoC1eodX5ftByMA4o1OPBbW8ABlTt5AgDB6jYOBsQARwwRRsTgY9zHPWk4C+N0BX1Y14NNWNAzdOhPVdaQLQRURI3BeQMHkNJ5EoVBFHgbA2ChGB-06Dd4FPMZz3odDMOw7E4gQCAqITeJ3Voxj3RY1BpBcNxQMyANHGDUM9EzPNIQLSzW3jfUkS4lEBOIs8L1QASiLRW970ErB4A3CoFFQWTlDIr0Yi3TtFFoLAaQwHysHocwwo1MCGnqY95kQYB4F4JgsWgJUf2gFhdkjDNAmzKwLisIA

### 💻 Code

```ts
export type Func<
  InnerArgs extends unknown[]
> = (this: Func<InnerArgs>) => void;

export type InferInnerArgs<T> = (
  T extends Func<infer InnerArgs> ? InnerArgs : never
);

function innerCall<InnerArgs extends unknown[]>(
  fn: (this: Func< InnerArgs>, ...args: InnerArgs) => void
) {
  // ...
}

const fn: Func<[number]> = function() {
  // infers Func<[number]>
  // this is correct
  this;

  // infers function innerCall<[id: number]>(fn: (this: Func<[id: number]>, id: number) => void): void
  // this is correct
  innerCall((id: number) => {
    // infers [number] - correct
    type argType = InferInnerArgs<typeof this>;
    // infers Func<[number]> - correct
    this;
  });

  // infers function innerCall<[id: string]>(fn: (this: Func<[id: string]>, id: string) => void): void
  // WRONG - thinks that the arrow fn's `this` type is a different one and doesn't report an error
  innerCall((id: string) => { // <- TS should require `id` to be a number
    // infers type argType = [number]
    // huh? here it understands that `this` should refer to a number arg
    type argType = InferInnerArgs<typeof this>;
    // infers this: Func<[number]>
    // the `this` type here is different than the `this` type in the signature inside `innerCall`, yet no error is reported
    this;
  });
};

```

### 🙁 Actual behavior

TypeScript knows that the `this` signature in the second `innerCall` differs from expected.

If you check the type for the second call, `this` refers to a string:
```
innerCall<[id: string]>(fn: (this: Func<[id: string]>, id: string) => void): void
```

However, inside the same function, `this` maps to:
```
this: Func<[number]>
```

This is an obvious mismatch that should trigger an error, but it doesn't.

### 🙂 Expected behavior

Type mismatch of `this` should be correctly reported as an error.
